File: backend/app/schemas/review.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.diff import DiffSummary
from app.schemas.eval import EvalResult, GuardrailStatus
from app.schemas.finding import Finding
from app.schemas.pr import ChangedFile, PRMetadata
from app.schemas.rag import ContextQualityResult, RAGQueryPlan, RetrievedContext
from app.schemas.risk import RiskSummary
from app.schemas.trace import TraceStep

RiskLevel = Literal["low", "medium", "high", "critical"]
Recommendation = Literal[
    "merge_ready",
    "merge_with_minor_fixes",
    "fix_before_merge",
    "do_not_merge",
]


class ReviewResult(BaseModel):
    request_id: str
    review_mode: str
    pr_url: str
    pr_metadata: PRMetadata
    changed_files: list[ChangedFile]
    diff_summary: DiffSummary
    risk_summary: RiskSummary
    rag_query_plan: RAGQueryPlan
    retrieved_context: list[RetrievedContext]
    context_quality: ContextQualityResult
    overall_score: int
    confidence: int
    risk_level: RiskLevel
    recommendation: Recommendation
    final_summary: str
    findings: list[Finding]
    guardrails: GuardrailStatus
    eval_result: EvalResult
    trace: list[TraceStep]

File: frontend/src/types/review.ts
export type Finding = {
  file_path: string;
  line_number?: number | null;
  severity: 'low' | 'medium' | 'high' | 'critical';
  issue: string;
  suggestion: string;
  pr_relevance_reason: string;
  relation_to_pr: string;
  evidence: string;
};

export type DiffSummary = {
  review_mode: string;
  main_change_type: string;
  main_area: string;
  goal_detected: boolean;
  requires_rag: boolean;
  required_review_types: string[];
};

export type TraceStep = {
  request_id: string;
  step_id: string;
  agent_id: string;
  status: string;
  summary: string;
};

export type ReviewResult = {
  request_id: string;
  review_mode: string;
  pr_url: string;
  diff_summary: DiffSummary;
  overall_score: number;
  confidence: number;
  risk_level: string;
  recommendation: string;
  final_summary: string;
  findings: Finding[];
  guardrails: {
    pr_scope_passed: boolean;
    evidence_passed: boolean;
    hallucination_passed: boolean;
    score_passed: boolean;
  };
  eval_result: {
    score: number;
    passed: boolean;
    reason: string;
  };
  trace: TraceStep[];
};
